workloads/imagenet/imagenet_jax/workload.py

```diff
@@ -0,0 +1,314 @@
+  # parameters.
+  def is_output_params(self, param_key: spec.ParameterKey) -> bool:
+    pass
+
+  def preprocess_for_train(
+      self,
+      selected_raw_input_batch: spec.Tensor,
+      selected_label_batch: spec.Tensor,
+      train_mean: spec.Tensor,
+      train_stddev: spec.Tensor,
+      rng: spec.RandomState) -> spec.Tensor:
+    return (selected_raw_input_batch, selected_label_batch)
+
+  def preprocess_for_eval(
+      self,
+      raw_input_batch: spec.Tensor,
+      raw_label_batch: spec.Tensor,
+      train_mean: spec.Tensor,
+      train_stddev: spec.Tensor) -> spec.Tensor:
+    return (raw_input_batch, raw_label_batch)
+
+  def create_model(self, *, model_cls, half_precision, **kwargs):
+    platform = jax.local_devices()[0].platform
+    if half_precision:
+      if platform == 'tpu':
+        model_dtype = jnp.bfloat16
+      else:
+        model_dtype = jnp.float16
+    else:
+      model_dtype = jnp.float32
+    return model_cls(num_classes=config.num_classes,
+                     dtype=model_dtype,
+                     **kwargs)
+
+  def initialized(self, key, image_size, model):
+    input_shape = (1, image_size, image_size, 3)
+    @jax.jit
+    def init(*args):
+      return model.init(*args)
+    variables = init({'params': key}, jnp.ones(input_shape, model.dtype))
+    model_state, params = variables.pop('params')
+    return params, model_state
+
+  _InitState = Tuple[spec.ParameterTree, spec.ModelAuxillaryState]
+  def init_model_fn(
+      self,
+      rng: spec.RandomState) -> _InitState:
+    model_cls = getattr(models, config.model)
+    model = self.create_model(
+      model_cls=model_cls, half_precision=config.half_precision)
+    params, model_state = self.initialized(rng, config.image_size, model)
+    self._param_shapes = jax.tree_map(
+      lambda x: spec.ShapeTuple(x.shape),
+      params)
+    learning_rate_fn = submission.create_learning_rate_fn(
+      config,
+      self.num_train_examples)
+
+    self.p_train_step = jax.pmap(
+      functools.partial(
+        submission.train_step,
+        model.apply,
+        learning_rate_fn=learning_rate_fn,
+        loss_fn=self.loss_fn,
+        loss_type=self.loss_type,
+        model_fn=self.model_fn,
+        compute_metrics=self.compute_metrics),
+      axis_name='batch')
+
+    self.p_eval_step = jax.pmap(
+      functools.partial(
+        submission.eval_step,
+        model.apply,
+        model_fn=self.model_fn,
+        compute_metrics=self.compute_metrics),
+      axis_name='batch')
+
+    state = TrainState(
+      step=0,
+      optimizer=None,
+      model_state=model_state)
+    return params, state
+
+  def model_fn(
+      self,
+      params: spec.ParameterTree,
+      augmented_and_preprocessed_input_batch: spec.Tensor,
+      model_state: spec.ModelAuxillaryState,
+      mode: spec.ForwardPassMode,
+      # rng: spec.RandomState, # TODO: Question— I'm not sure how to pass rng in
```

Comment:
that's a very good question! random state in a multidevice setting can be confusing (even moreso in a multihost, but we don't have to worry about that). in Jax we can just pass the same rng to each device, and then fold in the device index (example [here](https://github.com/google/init2winit/blob/master/init2winit/trainer.py#L151)) in order to get a per-device rng. we could also split the rng into `num_devices` rng's, and pass one to each device (but IIRC this is slower due to weird Jax issues)